### PR TITLE
[SCAF-3448] Nodejs script to update jira links

### PR DESCRIPTION
**Pull request by arsm100**
[SCAF-3448] Implement nodejs script to update Jira links.

[SCAF-3448] Add log files and summary.
* Add logger for INFO and ERROR logs.
* Save log output to log files.
* Print summary at the end of the execution.
* Add more details to README.md.

### Dry run summary:
Screenshot 2021-03-25 at 09 55 12

### Summary:
Screenshot 2021-03-25 at 09 54 11

### CODE REVIEW

**Comment by arsm100** [comment]
@jeromerodrigo-sr I can only add 1 reviewer, can you please check the settings?

File: index.js [hunk 1]
```diff
@@ -1,76 +1,119 @@
...
+    const authBuffer = Buffer.from(usernameAndPassword, 'utf-8');
+    requests.push(
+      axios.post(requestUrl, body, {
+        headers: {
+          'Authorization': `Basic ${authBuffer.toString('base64')}`,
+          'Content-Type': 'application/json'
+        }
+      })
+        .then(result => {
+          result.status === 201 ? createdLinksCount++ : preexistingLinksCount++;
```
Comment (jeromerodrigo-sr) [inline index.js:96]: Suggest to fix the indentation here

File: index.js [hunk 2]
```diff
@@ -1,76 +1,119 @@
...
-    .on("error", error => console.error(error))
-    .on("data", processRow)
-    .on("end", rowCount => console.log(`Parsed ${rowCount} rows`));
+  .pipe(csv.parse({headers: true}))
+  .on("error", console.error)
+  .on("data", processRow)
+  .on("end", async rowCount => {
+    log(INFO, `Parsed ${rowCount} rows`);
+    await Promise.allSettled(requests);
+    logSummary(rowCount, requests.length, createdLinksCount, preexistingLinksCount, errCount, isDryRun);
```
Comment (jeromerodrigo-sr) [inline index.js:117]: Q: If there are 100+ requests in the `requests` list, would this trigger all of them at the same time?

File: index.js [hunk 3]
```diff
@@ -1,76 +1,119 @@
...
+  .option("jira-auth", {
+    type: "string",
+    describe: "Username and password for the Jira user with 'Link Issues' permission in the following format 'username:password'"
+  })
+  .option("dry-run", {
+    type: "boolean",
+    default: false,
+    describe: "Script will not submit HTTP requests, if this flag is enabled."
+  })
+  .demandOption(["jiraUrl", "confUrl", "csv", "appId"])
```
Comment (jeromerodrigo-sr) [inline index.js]: I think we should add `jira-auth` as a mandatory argument and maybe fix the option names here e.g. `jiraUrl` -> `jira-url`

File: index.js [hunk 4]
```diff
@@ -1,76 +1,119 @@
...
+    if (isDryRun) {
+      log(INFO, `[DRY-RUN] Will update ${requestUrl}`);
+      return;
+    }
 
-    issueKeys.forEach(issueKey => {
-        const requestUrl = `${jiraEndpoint}/rest/api/latest/issue/${issueKey}/remotelink`;
-        console.log(requestUrl);
-    });
+    const authBuffer = Buffer.from(usernameAndPassword, 'utf-8');
```
Comment (jeromerodrigo-sr) [inline index.js]: The `usernameAndPassword` shouldn't change so I think it can be moved out of the `processRow` function.

File: package.json [hunk 5]
```diff
@@ -3,9 +3,11 @@
   "version": "1.0.0",
   "description": "Scripts to create Jira links for Scaffolding pages with missing links",
   "main": "index.js",
+  "type": "module",
   "author": "ServiceRocket",
   "license": "ISC",
```
Comment (jeromerodrigo-sr) [inline package.json:8]: nit: We should think about the licensing of the code since this is a public repo - There's a [site](https://choosealicense.com/) that helps with making the decision.

File: logger.js [hunk 6]
```diff
@@ -0,0 +1,58 @@
...
+        break;
+    }
+  }
+
+  static logSummary = (pageCount, issueCount, createdLinksCount, preexistingLinksCount, errCount, isDryRun) => {
+    const summary = `
+########################################################################################################################
+
+    ###############
+${isDryRun ? "    ##  DRY RUN  ##" : "    ###############"}
```
Comment (jeromerodrigo-sr) [inline logger.js:34]: super nit: Could add some indent here

**Review by jeromerodrigo-sr — CHANGES_REQUESTED** [review]

File: index.js (hunk 3 above)
Comment (arsm100) [reply index.js]: Addressed. Added `jira-auth` and fixed the naming to match the arguments. Somehow, it was still managing to detect missing arguments even before the rename, must have a camel case transformer or something similar built in 😄

File: index.js (hunk 4 above)
Comment (arsm100) [reply index.js]: Addressed.

File: index.js (hunk 1 above)
Comment (arsm100) [reply index.js:96]: Do you mean the `.then` block? I think it's commonly indented. In any case, I am using intelliJ's formatter and it's making it this way. Let me know if you were actually referring to something else.

Comment (jeromerodrigo-sr) [reply index.js:96]: Yeah, I'm looking at line 95 actually, the indentation looks out of place to me

File: index.js (hunk 3 above)
Comment (jeromerodrigo-sr) [reply index.js]: Yeah I think the library converts `camelCase` automatically :open_mouth:

File: index.js (hunk 1 above)
Comment (arsm100) [reply index.js:96]: I still can't see it lol.

line 90 has 3 indents and line 95 closes it and also has 3 indents. Not sure what I am missing 🕵️‍♂️

Comment (jeromerodrigo-sr) [reply index.js:96]: We can compare with line 99-100 which is what I usually expect. Anyway no strong opinion on this one, it's just indentation

File: index.js (hunk 2 above)
Comment (arsm100) [reply index.js:117]: Not exactly. Each request will be processed as soon as it's ready and node can allocate resources to it. To break it down:

1. The axios request is created and assigned callbacks.
2. The request (with its callbacks) is added to the `requests` array.
3. As requests are sent and their responses become available, their callbacks are triggered (while other requests are still being queued).
4. Once all requests have been queued (some will have been completed), line 117 will await for all the requests to be settled before printing the summary.

In short, requests are processed as soon as they are ready, they don't actually wait until here to start triggering.

File: logger.js (hunk 6 above)
Comment (arsm100) [reply logger.js:34]: Since this is a multiline string, any added indent would actually be printed breaking the alignment of the title box.

File: index.js (hunk 2 above)
Comment (jeromerodrigo-sr) [reply index.js:117]: Thanks for the explanation, my only concern is the number of HTTP requests being sent out at one time - do you see a scenario where a large number of HTTP requests would overload the server?

File: index.js (hunk 1 above)
Comment (arsm100) [reply index.js:96]: Actually, line 100 has the same indentation as line 96.
Screenshot 2021-03-25 at 12 09 41

I think it gets confusing because intelliJ by default does this:
```js
axios.post(url,{
  headers: {}
})
  .then()
  .catch()
```
instead of:
```js
axios.post(url,{
  headers: {}
})
.then()
.catch()
```

File: index.js (hunk 2 above)
Comment (arsm100) [reply index.js:117]: I think that the server will be okay with these lightweight requests as long as they're not executed during high traffic when the server is already taxed. Honestly, if I had a lot of time, I would have introduced batching to be extra safe but it's just a nice to have.
For reference, I ran 600+ requests on my dockerised Jira server which is running on my modestly-speced machine together with Confluence, Postgres and pgAdmin and the server did ok to handle it.
I am more concerned about the protections that may be present to prevent DoS which why I added a note about that in the README.md.

**Review by jeromerodrigo-sr — APPROVED** [review]

File: package.json (hunk 5 above)
Comment (arsm100) [reply package.json:8]: I think `ISC` should be ok for public projects. Full license text [here](https://choosealicense.com/licenses/isc/).

File: index.js [hunk 7]
```diff
@@ -1,76 +1,119 @@
...
-    });
+    requests.push(
+      axios.post(requestUrl, body, {
+        headers: {
+          'Authorization': `Basic ${authBuffer.toString('base64')}`,
+          'Content-Type': 'application/json'
+        }
+      })
+        .then(result => {
+          result.status === 201 ? createdLinksCount++ : preexistingLinksCount++;
```
Comment (arsm100) [inline index.js:96]: #ShareTheKnowledge

Working on this helped me better understand how incrementing this way in an asynchronous setting is valid in JavaScript when in Java you'd need to use an [AtomicInteger](https://docs.oracle.com/javase/8/docs/api/java/util/concurrent/atomic/AtomicInteger.html#incrementAndGet--).

This [answer](https://stackoverflow.com/a/7266985) also has a nice short summary.

**Review by 8gb — APPROVED** [review]

File: index.js [hunk 8]
```diff
@@ -1,76 +1,119 @@
...
+          log(INFO, `Update success: ${requestUrl} => ${result.status} ${JSON.stringify(result.data)}`);
```
Comment (8gb) [inline index.js:97]: Question: Do we need to add `Promise.resolve()`?

Comment (arsm100) [reply index.js:97]: No need to resolve here because this is executed in the callback after the request has been completed successfully.

File: package.json (hunk 5 above)
Comment (jeromerodrigo-sr) [reply package.json:8]: super nit: Perhaps we can add the LICENSE file to the repo :sunglasses:

File: index.js [hunk 9]
```diff
@@ -1,76 +1,119 @@
...
-    .pipe(csv.parse({ headers: true }))
-    .on("error", error => console.error(error))
-    .on("data", processRow)
-    .on("end", rowCount => console.log(`Parsed ${rowCount} rows`));
+  .pipe(csv.parse({headers: true}))
+  .on("error", console.error)
+  .on("data", processRow)
+  .on("end", async rowCount => {
+    log(INFO, `Parsed ${rowCount} rows`);
+    await Promise.allSettled(requests);
```
Comment (afifsohaili-sr) [inline index.js:116]: nit: Since you're using `allSettled` and you're getting a list of `resolved` and `rejected` promises at the end, technically you can just count the number of rejected promises and not have to keep your own `errCount`.

See https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise/allSettled#using_promise.allsettled.

Not blocking the PR approval

File: extract-jira-link-pages.sql [hunk 10]
```diff
@@ -1,6 +1,22 @@
...
+		os_propertyentry op
+	where
+		op.entity_key like '~metadata.%'
+        and op.text_val like '%structured-macro ac:name=&quot;jira&quot;%'
+	group by
+		entity_id) op3 on
+	op2.entity_id = op3.entity_id
+	and op2.entity_key = op3.entity_key
+inner join "content" c on
+	c.contentid = op2.entity_id
```
Comment (afifsohaili-sr) [inline extract-jira-link-pages.sql:22]: Are we telling the customer about `limit` and `offset`? I don't see that in the README

**Review by ainzizou — APPROVED** [review]

File: index.js [hunk 11]
```diff
@@ -1,76 +1,119 @@
 #!/usr/bin/env node
 
-const fs = require('fs');
-const path = require('path');
-const csv = require('fast-csv');
+import fs from 'fs';
+import path from 'path';
+import axios from 'axios';
+import csv from 'fast-csv';
```
Comment (mbbarrry) [inline index.js]: empty line

**Review by mbbarrry — APPROVED** [review]

File: index.js (hunk 9 above)
Comment (arsm100) [reply index.js:116]: This suggestion would only be possible if I were saving promises from axios to the array and then using `Promise.allSettled` to resolve them at one go. However, In this case, each axios request has its own callbacks and `.catch` block. Because of this local catching, all promises return `fulfilled`. Therefore, the `rejected` count is always 0.

I initially wrote the success and failure callbacks here based on the `fulfilled` or `rejected` status but then moved it back locally to each request to be able to log more context about the process being executed. As an example, the `rejected` reason looks like this making it impossible to know which request of the possibly hundreds this originated from
Screenshot 2021-03-26 at 12 56 50

File: extract-jira-link-pages.sql (hunk 10 above)
Comment (arsm100) [reply extract-jira-link-pages.sql:22]: Thanks for the reminder, I'll add that to the README.

File: index.js (hunk 11 above)
Comment (arsm100) [reply index.js]: Addressed

**Review by afifsohaili-sr — APPROVED** [review]
